src/modules/python/pbs/v1/_pmi_cray.py

```diff
@@ -0,0 +1,424 @@
+# coding: utf-8
+"""
+
+/*
+#  Copyright (C) 1994-2017 Altair Engineering, Inc.
+#  For more information, contact Altair at www.altair.com.
+#
+#  This file is part of the PBS Professional ("PBS Pro") software.
+#
+#  Open Source License Information:
+#
+#  PBS Pro is free software. You can redistribute it and/or modify it under the
+#  terms of the GNU Affero General Public License as published by the Free
+#  Software Foundation, either version 3 of the License, or (at your option) any
+#  later version.
+#
+#  PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+#  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+#  PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
+#
+#  You should have received a copy of the GNU Affero General Public License along
+#  with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#  Commercial License Information:
+#
+#  The PBS Pro software is licensed under the terms of the GNU Affero General
+#  Public License agreement ("AGPL"), except where a separate commercial license
+#  agreement for PBS Pro version 14 or later has been executed in writing with Altair.
+#
+#  Altair’s dual-license business model allows companies, individuals, and
+#  organizations to create proprietary derivative works of PBS Pro and distribute
+#  them - whether embedded or bundled with other software - under a commercial
+#  license agreement.
+#
+#  Use of Altair’s trademarks, including but not limited to "PBS™",
+#  "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+#  trademark licensing policies.
+ *
+ */
+"""
+__doc__ = """
+This module is used for Cray systems.
+"""
+
+import pbs
+import os
+import sys
+import stat
+from pbs.v1._pmi_types import BackendError
+from pbs.v1._pmi_utils import _running_excl, _pbs_conf, _get_vnode_names, \
+    _svr_vnode
+
+pbsexec = _pbs_conf("PBS_EXEC")
+if pbsexec is None:
+    raise BackendError("PBS_EXEC not found")
+
+sys.path.append(os.path.join(pbsexec, "python", "lib", "python2.7"))
+sys.path.append(os.path.join(pbsexec, "python", "lib", "python2.7",
+                "lib-dynload"))
+
+
+def launch(id, args):
+    """
+    Run capmc and return the structured output.
+    """
+    import json
+
+    # full path to capmc given by Cray
+    cmd = "/opt/cray/capmc/default/bin/capmc"
```

Comment:
Use os.path.join() to form capmc "cmd".

Reply:
I do not see the need to use join() unless we are making a path which could be same on Linux and Windows. In this case it is only Linux and even if we have to support this feature on windows it will be different path. Can you please tell me anything that can cause issues of using path like this?

Reply:
The comment from @suresh-thelkar is valid. While what you have will work fine on Linux systems, we use os.path.join() for consistency and portability. Maybe this code will never run on Windows, but if it ever needs porting that will be one less issue to address.

Reply:
Fixed it.